### PR TITLE
feat: add Database Operation tool

### PR DESCRIPTION
**Pull request by XiaoLey**
# Description

Introduced a new Database Operation tool that allows execution of SQL statements and returns corresponding results. This enhancement improves the convenience of database operations.

Currently supported ORDBMS include **PostgreSQL** and **MySQL**.

image

image

image

https://github.com/langgenius/dify/assets/46273173/6bb2e389-b6af-4a11-a7e0-884d87e4f26d

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] TODO

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [ ] `optional` I have added tests that prove my fix is effective or that my feature works
- [ ] `optional` New and existing unit tests pass locally with my changes

### CODE REVIEW

File: api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.yaml [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    label:
+      en_US: Username
+      zh_Hans: 用户名
+    human_description:
+      en_US: Username
+      zh_Hans: 用户名
+    llm_description: Database username
+    form: form
+  - name: password
+    type: string
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.yaml]: Use `secret-input` in type, which helps to mask part of the value on frontend side.
Please refer to exsiting tools for the proper usage.

File: api/core/tools/provider/builtin/novitaai/tools/novitaai_modelquery.py [hunk 2]
```diff
@@ -42,17 +42,18 @@ def _invoke(self,
...
+            sd_name_array = [model['sd_name_in_api'] for model in models_data] if len(models_data) > 0 else []
             result_str = json.dumps(sd_name_array)
         elif result_type == 'name array':
-            name_array = [model['name'] for model in models_data]
+            name_array = [model['name'] for model in models_data] if len(models_data) > 0 else []
             result_str = json.dumps(name_array)
         elif result_type == 'name sd_name pair array':
-            name_sd_name_pair_array = [{'name': model['name'], 'sd_name': model['sd_name_in_api']} for model in models_data]
+            name_sd_name_pair_array = [{'name': model['name'], 'sd_name': model['sd_name_in_api']}
+                                       for model in models_data] if len(models_data) > 0 else []
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/novitaai/tools/novitaai_modelquery.py]: If the novita ai tool is not part of the newly introduced tool for db query, please remain files in novitaai/tools folder untouched.

File: api/core/tools/provider/builtin/databaseoperation/databaseoperation.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from core.tools.errors import ToolProviderCredentialValidationError
+from core.tools.provider.builtin_tool_provider import BuiltinToolProviderController
+
+
+class DatabaseOperationProvider(BuiltinToolProviderController):
+    def _validate_credentials(self, credentials: dict) -> None:
+        try:
+            pass
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/databaseoperation/databaseoperation.py]: At least init a tool instance with no args anyway, making sure the tool class is loadable?

File: api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        password = tool_parameters.get('password')
+        dbname = tool_parameters.get('dbname')
+
+        result_str = ''
+        if dbtype == 'postgresql':
+            result_str = self._pg_exec(sql, output_format, host, port, username, password, dbname)
+        elif dbtype == 'mysql':
+            result_str = self._mysql_exec(sql, output_format, host, port, username, password, dbname)
+        else:
+            raise ValueError(f"Unsupported database type: {dbtype}")
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.py]: Please add the supported database types in the PR description.

**Review by bowenliang123 — COMMENTED** [review]
A great tool for fetching data from RMDB. Great work.
Some modification suggestions from me.

**Comment by XiaoLey** [comment]
review please @bowenliang123

File: api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                result_list = [dict(zip(fields_list, data_tuple)) for data_tuple in result_data_list]
+                if output_format == 'markdown':
+                    result_str = self._convert_to_markdown_table(fields_list, result_list)
+                else:
+                    result_str = json.dumps(result_list)
+        else:
+            result_str = ''
+
+        cur.close()
+        conn.close()
```
Comment (bowenliang123) [inline api/core/tools/provider/builtin/databaseoperation/tools/databasecontrol_sqlexec.py]: Make sure the opened resources are properly closed. Use `with` or `try...finally` instead.

**Comment by bowenliang123** [comment]
I would also suggest to consider apply some design patterns, like Template Method.
Avoid implementing the whole process for handling MySQL and PG queries from scratch.

**Comment by bowenliang123** [comment]
And there's a CRITICAL security issue to consider.
Letting the LLM to decide the whole SQL may cause huge injection leak.
I would suggest limited the types of SQL into query first, only allowing `SELECT` prefix sqls.
And for other DML and DDL, they can be covered in future, and carefully design proper handling policies.

**Comment by XiaoLey** [comment]
@bowenliang123 
I apologize for the oversight. When designing this tool, my primary focus was on the need to record data, as sometimes it's important to have the results of the current process for future reference. Your feedback made me realize that this tool is more complex than I initially thought. I will carefully redesign it to ensure it has good versatility. For now, I will not open all database operation features at once until I have perfected the handling of various exceptions. If you have any further comments or suggestions, please feel free to let me know. I will fully cooperate and hope that this tool can meet as many needs as possible. Thank you for your valuable input!

**Pull request closed without merging**